patterns/strategy.md

```diff
@@ -54,29 +56,25 @@ impl Report {
 
 struct Text;
 impl Formatter for Text {
-    fn format(&self, data: &Data, s: &mut String) {
-        *s = data
-            .iter()
-            .map(|(key, val)| format!("{} {}\n", key, val))
-            .collect();
+    fn format(&self, data: &Data, buf: &mut String) {
+        for (k, v) in data {
+            let entry = format!("{} {}\n", k, v);
+            buf.push_str(&entry);
+        }
     }
 }
 
 struct Json;
 impl Formatter for Json {
-    fn format(&self, data: &Data, s: &mut String) {
-        *s = String::from("[");
-        let mut iter = data.into_iter();
-        if let Some((key, val)) = iter.next() {
-            let entry = format!(r#"{{"{}":"{}"}}"#, key, val);
-            s.push_str(&entry);
-            while let Some((key, val)) = iter.next() {
-                s.push(',');
-                let entry = format!(r#"{{"{}":"{}"}}"#, key, val);
-                s.push_str(&entry);
-            }
+    fn format(&self, data: &Data, buf: &mut String) {
+        buf.push('[');
+        for (k, v) in data.into_iter() {
+            let entry = format!(r#"{{"{}":"{}"}}"#, k, v);
+            buf.push_str(&entry);
+            buf.push(',');
         }
-        s.push(']');
+        buf.pop(); // remove extra , at the end
+        buf.push(']');
```

Comment:
I was wondering if we could do replace but I don't think there is a safe and easy to do it.